live-build/misc/upgrade-scripts/upgrade

```diff
@@ -0,0 +1,167 @@
+
+	exit 2
+}
+
+function cleanup_container() {
+	#
+	# If the CONTAINER variable is empty, this means container
+	# creation failed. In that case, the container creation process
+	# will have already cleaned up after itself, and there's no
+	# further cleanup for us to do here.
+	#
+	[[ -z "$CONTAINER" ]] && return
+
+	"$IMAGE_PATH/upgrade-container" stop "$CONTAINER" ||
+		warn "failed to stop container '$CONTAINER'"
+
+	"$IMAGE_PATH/upgrade-container" destroy "$CONTAINER" ||
+		warn "failed to stop container '$CONTAINER'"
+}
+
+function cleanup_in_place_upgrade() {
+	#
+	# Capture the exit code here, and use that to determine if
+	# upgrade verification was successful or not.
+	#
+	local rc="$?"
+
+	if [[ "$rc" == "0" || "$DLPX_DEBUG" != "true" ]]; then
+		#
+		# On success, or on failure when DLPX_DEBUG is not true,
+		# we clean up the container previously created. When
+		# DLPX_DEBUG is true, we leave the container around on
+		# failure to aid debugging efforts.
+		#
+		cleanup_container
+	fi
+
+	return "$rc"
+}
+
+function upgrade_in_place() {
+	trap cleanup_in_place_upgrade EXIT
+
+	CONTAINER=$("$IMAGE_PATH/upgrade-container" create in-place)
+	[[ -n "$CONTAINER" ]] || die "failed to create container"
+
+	"$IMAGE_PATH/upgrade-container" start "$CONTAINER" ||
+		die "failed to start container '$CONTAINER'"
+
+	"$IMAGE_PATH/upgrade-container" run "$CONTAINER" \
+		"$IMAGE_PATH/execute" ||
+		die "'$IMAGE_PATH/execute' failed in container '$CONTAINER'"
+
+	"$IMAGE_PATH/upgrade-container" run "$CONTAINER" \
+		"$IMAGE_PATH/verify" "$@" ||
+		die "'$IMAGE_PATH/verify' '$*' failed in container '$CONTAINER'"
+
+	"$IMAGE_PATH/execute" ||
+		die "'$IMAGE_PATH/execute' failed in running appliance."
+}
+
+function cleanup_not_in_place_upgrade() {
+	#
+	# Capture the exit code here, and use that to determine if
+	# upgrade verification was successful or not.
+	#
+	local rc="$?"
+
+	if [[ "$rc" != "0" && "$DLPX_DEBUG" != "true" ]]; then
+		#
+		# On failure and when DLPX_DEBUG is not true, we clean
+		# up the container previously created.
+		#
+		# Unlike in-place upgrades, the container is not cleaned
+		# up on success; instead the container will have been
+		# already stopped and converted to be used as the root
+		# filesystem for the appliance after the next reboot.
+		#
+		cleanup_container
+	fi
+
+	return "$rc"
+}
+
+function upgrade_not_in_place() {
+	trap cleanup_not_in_place_upgrade EXIT
+
+	CONTAINER=$("$IMAGE_PATH/upgrade-container" create not-in-place)
+	[[ -n "$CONTAINER" ]] || die "failed to create upgrade container"
+
+	"$IMAGE_PATH/upgrade-container" start "$CONTAINER" ||
+		die "failed to start upgrade container '$CONTAINER'"
+
+	"$IMAGE_PATH/upgrade-container" run "$CONTAINER" \
+		"$IMAGE_PATH/execute" ||
+		die "'$IMAGE_PATH/execute' failed in container '$CONTAINER'"
+
+	"$IMAGE_PATH/upgrade-container" run "$CONTAINER" \
+		"$IMAGE_PATH/verify" "$@" ||
+		die "'$IMAGE_PATH/verify' '$*' failed in container '$CONTAINER'"
```

Comment:
I don't think we would want to split the code for `verify` like this, the app-gate code would break
The app-gate wants to be able to just simply call a `verify` script that will take care of setting up all the container stuff and running the checks.  It looks like with this change we no longer have that capability anymore.

How about doing something like renaming your new `verify` to a different script?  And then use that as a building block for the old `verify` script and your new `upgrade` script.  That way we can keep this refactoring while maintaining the old functionality?

Reply:
Sure. We could do something like that, but we also need to think through how this will work for “not-in-place” upgrades and verification. 

I can definitely do what you suggest to avoid breaking the current app-gate code, but let’s discuss (perhaps at the next upgrade meeting) how we want this to work long term, considering both “in-place” and “not-in-place” upgrades; I think we want to be able to run the same verification logic is both cases.

For now, I’ll do as you suggest.

Reply:
OK, I've created a new `verify-impl` script that's used by the new `upgrade` script; `verify` will create/destroy the verification container as needed, just like prior to this change. Let me know what you think when you get a chance to give this another look.